Improve performance of Unit division

exupero opened the pull request.

Commit:
Check equality at beginning of getQuotientInstance

oehme commented on src/main/java/javax/measure/unit/ProductUnit.java:
```diff
@@ -188,6 +188,9 @@ static Unit<? extends Quantity> getProductInstance(Unit<?> left, Unit<?> right)
      */
     static Unit<? extends Quantity> getQuotientInstance(Unit<?> left, Unit<?> right) {
         Element[] leftElems;
+        if (left == right) {
```
We should check equals, not identity.

afbarbaro replied:
Checking identity here should be good enough, right? since we now cache instances of `Unit` subclasses.

oehme replied:
We only cache ProductUnit, none of the others. But as mentioned further in the PR, this doesn't solve the issue and there's a better solution.

oehme reviewed (commented):
Do you have some benchmark results?

oehme commented:
> There seem to be much bigger gains from just using doGetInstance within getQuotientInstance:

We always need to go through the cache when creating new units, so we get the same unit inside a SignalSegment, rather than a new unit instance for each Sample. 

The problem with this benchmark is that it only shows you the cost of the initial division, but never uses that unit again for any follow-up computations. So you get a skewed perception of it. That said, I think we can make this faster:

I [profiled](https://github.com/seeq12/jscience/files/9434256/divideby.html.zip) this scenario and the bottleneck is `Unit.inverse`, which I'm sure we can improve. One solution would be to remember the inverse of each unit in a field, just like we did for the `standardUnit`.

oehme commented:
Also, I'm not sure why we're hitting the `inverse` code path here at all. In the SAME_UNITS case we should have `leftUom==rightUom`, but that's not happening. Might be a problem in the benchmark setup.

oehme commented:
Okay, I understand the problem with the benchmark now. We first divide e.g. `m/m`. That gives us `Unit.ONE`. Then we divide that by `m` again, which results in the `unit.inverse` call, with the result `1/m`.

oehme committed:
Cache the inverse of each Unit

This speeds up divisions significantly.

oehme committed:
Speed up Product/TransformedUnit comparisons

Make use of their precomputed hashCode in equals checks
before moving on to the more expensive checks.

oehme commented:
I've pushed up changes that make this a lot faster. Please have a look @exupero

oehme committed:
Cache the hashCode for base units too

We keep these base units in constant fields,
so they are reused all the time and benefit
from this optimization.

exupero commented:
Nice finds, that made a big improvement:
![benchmarks](https://user-images.githubusercontent.com/287159/186959176-e9db5bb5-9a29-4773-85bb-c9517a6ca826.png)

oehme reviewed (approved):
Thanks for checking, this is good to go from my side then.

exupero commented:
@afbarbaro any comments before I merge?

afbarbaro commented:
> @afbarbaro any comments before I merge?

Sorry I didn't get to this before - I reviewed and it LGTM!
